tests/repo_finder/test_provenance_extractor.py

```diff
@@ -0,0 +1,473 @@
+    attestations = json_extract(witness_github_provenance, ["predicate", "attestations"], list)
+    assert _json_modify(attestations[attestation_index], keys, new_value)
+    with pytest.raises(ProvenanceExtractionException):
+        _test_extract_repo_and_commit_from_provenance(witness_github_provenance)
+
+
+def test_witness_github_remove_attestation(witness_github_provenance: dict[str, JsonType]) -> None:
+    """Test removing Git attestation from Witness V0.1 GitHub provenance."""
+    attestations = json_extract(witness_github_provenance, ["predicate", "attestations"], list)
+    assert _json_modify(witness_github_provenance, ["predicate", "attestations"], attestations[:1])
+    with pytest.raises(ProvenanceExtractionException):
+        _test_extract_repo_and_commit_from_provenance(witness_github_provenance)
+
+
+@pytest.mark.parametrize(
+    ("type_", "predicate_type"),
+    [
+        ("https://in-toto.io/Statement/v0.1", "https://slsa.dev/provenance/v1"),
+        ("https://in-toto.io/Statement/v1", "https://slsa.dev/provenance/v0.2"),
+        ("https://in-toto.io/Statement/v1", "https://slsa.dev/provenance/v0.1"),
+        ("https://in-toto.io/Statement/v1", "https://witness.testifysec.com/attestation-collection/v0.1"),
+    ],
+)
+def test_invalid_type_payloads(type_: str, predicate_type: str) -> None:
+    """Test payloads with invalid type combinations."""
+    payload: dict[str, JsonType] = {"_type": type_, "predicateType": predicate_type, "subject": [], "predicate": {}}
+    with pytest.raises(ProvenanceExtractionException):
+        _test_extract_repo_and_commit_from_provenance(payload)
+
+
+def _test_extract_repo_and_commit_from_provenance(
+    payload: dict[str, JsonType], expected_repo: str = "", expected_commit: str = ""
+) -> None:
+    """Accept a provenance and extraction function, assert the extracted values match the expected ones."""
+    provenance = validate_intoto_payload(payload)
+    repo, commit = extract_repo_and_commit_from_provenance(provenance)
+    assert expected_repo == repo
+    assert expected_commit == commit
+
+
+def _json_modify(entry: JsonType, keys: list[str], new_value: JsonType) -> bool:
+    """Modify the value found by following the list of depth-sequential keys inside the passed JSON dictionary.
+
+    The found value will be overwritten by the new_value parameter.
+    If new_value is None, the value will be removed.
+    If the final key does not exist, it will be created as new_value.
+    """
+    target = entry
+    last_target = None
+
+    for key in keys:
+        if not isinstance(target, dict):
+            return False
+        if key not in target:
+            return False
+        last_target = target
+        target = target[key]
+
+    if last_target is None:
+        return False
+
+    if new_value is None:
+        del last_target[keys[len(keys) - 1]]
+    else:
+        last_target[keys[len(keys) - 1]] = new_value
+
+    return True
```

Comment:
I would personally do it like this.

```suggestion
def _json_modify(entry: JsonType, keys: list[str], new_value: JsonType) -> None:
    """Modify the value found by following the list of depth-sequential keys inside the passed JSON dictionary.

    The found value will be overwritten by the `new_value` parameter.
    If `new_value` is `None`, the value will be removed.
    If the final key does not exist, it will be created as `new_value`.
    """
    target = entry

    for key in keys[:-1]:
        target = target[key]

    if new_value is None:
        del target[keys[-1]]
    else:
        target[keys[-1]] = new_value
```

No reason to do any validation and return `bool`. Just let the test raise errors and give us the stack trace.

Reply:
I see where you are coming from, but this would require disabling multiple `mypy` checks to be usable.

Reply:
I've made use of the `json_extract` function here instead.

Reply:
I assume it should be OK if we don't type-annotate `entry` as `JsonType`, which is too wide of a type in this context because test input must always be nested dictionaries anyway? Either annotating it as `dict` or even `Any` is fine for me here.